### ISSUE TITLE
[mqtt.homie] Binding subscribes/publishes with QoS 0 although openHAB setting is different

### ISSUE THREAD

**Issue by stedon81** [openhab/openhab-addons#9919]
## Expected Behavior
The binding should subscribe at the broker with the QoS level defined in the MQTT-Broker-Thing's settings.
If the binding connects to the broker and subscribes for the channels of a homie thing, it should specify the QoS level defined by the user.
Likewise, the binding should publish messages to the broker with the QoS level defined in the MQTT-Broker settings.

## Current Behavior
If the binding connects to the broker and subscribes for the channels of a homie thing, it always subscribes with QoS 0, although my settings in openHAB say that QoS 2 should be used.

DEBUG log from mosquitto broker:
```
1611354225: Received SUBSCRIBE from openHAB
1611354225:     homie/T_KG_Tuer_Fingerabdrucksensor/fingerprint/touch (QoS 0)
1611354225: openHAB 0 homie/T_KG_Tuer_Fingerabdrucksensor/fingerprint/touch
1611354225: Sending SUBACK to openHAB
1611354225: Sending PUBLISH to openHAB (d0, q0, r1, m0, 'homie/T_KG_Tuer_Fingerabdrucksensor/fingerprint/touch', ... (5 bytes))
```
Also, if the binding publishes a value to the broker, it publishes it with QoS 0:

```
Received PUBLISH from openHAB (d0, q0, r0, m0, 'homie/T_KG_Tuer_Fingerabdrucksensor/bell/response/set', ... (1 bytes))
```

Broker thing configuration:
```
UID: mqtt:broker:mosquitto
label: GTW_Mosquitto
thingTypeUID: mqtt:broker
configuration:
  publickeypin: false
  lwtQos: 1
  lastwill_qos: 1
  clientid: ce05f2fc-7936-4193-ab5a-f5e87258aa60
  keepAlive: 60
  clientID: openHAB
  retain: true
  secure: false
  certificatepin: false
  keep_alive_time: 60000
  password: *****************
  qos: 2
  reconnectTime: 60000
  port: 1884
  host: 127.0.0.1
  lwtRetain: true
  reconnect_time: 60000
  username: ****************
  enableDiscovery: true
```

## Possible Solution
No idea. Seems strange, because `createBrokerConnection` sets the QoS from configuration in the connection object.
Connection's `subscribeRaw` passes the qos to the mqttClient.
Suspicion:
In `Mqtt3AsyncClientWrapper.subscribe` as well as in `Mqtt3AsyncClientWrapper.publish`, a function `getMqttQosFromInt` tries to convert the `int` parameter `qos` to an enum. This function returns the wrong QoS enum values for qos==0 and qos==1, but the correct one for qos==2. So this also is a red herring. I can't explain to myself why the broker only sees requests for QoS 0.

## Your Environment
* Version used: openHABian 3.0.0-release
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...): Debian

**Comment by stedon81** [comment]
I posted the wrong mappings of function `getMqttQosFromInt` as a bug to openhab-core:
https://github.com/openhab/openhab-core/issues/2145
This bug here addresses the issue that the QoS levels are not set at all.

**Comment by jochen314** [comment]
This looks like a but in the core to me...

in <code>org.openhab.core.io.transport.mqtt.internal.client.MqttAsyncClientWrapper</code> we have:
```
    protected MqttQos getMqttQosFromInt(int qos) {
        switch (qos) {
            case 0:
                return MqttQos.AT_LEAST_ONCE;
            case 1:
                return MqttQos.AT_MOST_ONCE;
            case 2:
                return MqttQos.EXACTLY_ONCE;
            default:
                throw new IllegalArgumentException("QoS needs to be 0, 1 or 2.");
        }
    }
```
But http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718099 has:
 

>  4.3.1 QoS 0: At most once delivery
>   4.3.2 QoS 1: At least once delivery
>   4.3.3 QoS 2: Exactly once delivery

The mapping for 0 and 1 are swapped.

**Comment by jochen314** [comment]
I should have read the other comments....

**Comment by jochen314** [comment]
The homie convention states (https://homieiot.github.io/specification/#qos-and-retained-messages):
```
2.3 QoS and retained messages
The nature of the Homie convention makes it safe about duplicate messages, 
so the recommended QoS for reliability is At least once (1). 
...
```
That is the reason, why the code ties the QoS to 1.

**Comment by stedon81** [comment]
Hi jochen 314,

I crawled the most recent code source on GitHub for quite a long time and haven't found a location where homie ties this to QoS 1. I agree that this is true for older versions of the code, however, the most recent one seems to just use the QoS from the openhab config.
Can you please:
1) point us to the code location where this QoS 1 setting happens (in that case, this bug can be closed and the bug in core must just be fixed)
2) tell which code version is active in openHAB 3.0.0-release (I'm not familiar enough with github to find out)

Thanks,
stedon81

**Comment by jochen314** [comment]
https://github.com/openhab/openhab-addons/blob/f5ee68555689c6ea44b07bb02bcc1b33aba8656f/bundles/org.openhab.binding.mqtt.homie/src/main/java/org/openhab/binding/mqtt/homie/internal/handler/HomieThingHandler.java#L127
```
protected CompletableFuture<@Nullable Void> start(MqttBrokerConnection connection) {
        logger.debug("About to start Homie device {}", device.attributes.name);
        if (connection.getQos() != 1) {
            // QoS 1 is required.
            logger.warn(
                    "Homie devices require QoS 1 but Qos 0/2 is configured. Using override. Please check the configuration");
            connection.setQos(1);
        }
```

**Comment by stedon81** [comment]
Hi jochen314,
you're right, I even stumbled over this code location and saw the check and the warn message, but didn't really see that it is also set there. OK, with that, I think that this bug will be solved as soon as openhab/openhab-core#2145 is solved.
Regards,
stedon81

**Comment by kaikreuzer** [comment]
Fixed by https://github.com/openhab/openhab-core/pull/2208.
